rtslib/root.py

```diff
@@ -290,17 +290,89 @@ def so_err_func(x):
         '''
         if not save_file:
             save_file = default_save_file
 
+        if block_name:
+            current = self.dump()
+
+            with open(save_file, "r") as f:
+                saveconf = json.loads(f.read())
+                f.close()
+
+            fetch_cur_so = False
+            fetch_cur_tg = False
+            # Get the given block current storageObj configuration
+            for sidx, sobj in enumerate(current.get('storage_objects', [])):
+                if sobj['name'] == block_name:
+                    current_so = current['storage_objects'][sidx]
+                    fetch_cur_so = True
+                    break
+
+            # Get the given block current target configuration
+            if fetch_cur_so:
+                for tidx, tobj in enumerate(current.get('targets', [])):
+                    if fetch_cur_tg:
+                        break
+                    for luns in tobj.get('tpgs', []):
+                        if fetch_cur_tg:
+                            break
+                        for lun in luns.get('luns', []):
+                            if os.path.basename(lun['storage_object']) == block_name:
+                                current_tg = current['targets'][tidx]
+                                fetch_cur_tg = True
+                                break
+
+            fetch_sav_so = False
+            fetch_sav_tg = False
+            # Get the given block storageObj from saved configuration
+            for sidx, sobj in enumerate(saveconf.get('storage_objects', [])):
+                if sobj['name'] == block_name:
+                    # Merge StorageObj
+                    if fetch_cur_so:
+                        saveconf['storage_objects'][sidx] = current_so;
+                    # Remove StorageObj
+                    else:
+                        saveconf['storage_objects'].remove(saveconf['storage_objects'][sidx])
+                    fetch_sav_so = True
+                    break
+
+            # Get the given block target from saved configuration
+            if fetch_sav_so:
+                for tidx, tobj in enumerate(saveconf.get('targets', [])):
+                    if fetch_sav_tg:
+                        break
+                    for luns in tobj.get('tpgs', []):
+                        if fetch_sav_tg:
+                            break
+                        for lun in luns.get('luns', []):
+                            if os.path.basename(lun['storage_object']) == block_name:
+                                # Merge target
+                                if fetch_cur_tg:
+                                    saveconf['targets'][tidx] = current_tg;
+                                # Remove target
+                                else:
+                                    saveconf['targets'].remove(saveconf['targets'][tidx])
+                                fetch_sav_tg = True
+                                break
```

Comment:
This together with the targetcli-fb's patch test okay for me till now and will test it more later.

Just one suggestion:
Since this is the same as the code of searching the current so/target, how about let's put it in a function ? This maybe more readable.